### DISCUSSION TITLE
Add the non standard hardware supported by the new XT machines

### DISCUSSION THREAD

**Discussion by Fenix770** [86Box/86Box#4166] [Ideas]
**Is your feature request related to a problem? Please describe.**

The recently added machines, supports some hardware that is not standard. I know it will take time and the specific datasheet to implement, but it's very interesting, since not all XT machines support them, and also from a historic point of view. 
May be they can be implemented sometime in the future.

Machines

1. Philiphs P3105

This machine comes with ATI Small Wonder video card that support both Hercules and CGA graphics and also supports
the Plantronics Colorplus video modes. But it only uses  monochrome 
shades of green when a color CGA monitor is used.

I found the specific ATI drivers for this video card/chip, but they are in Dutch.
The chipset is based on ATI 18700 according this site:

Here Is the drivers I found: 
[Philips P3105.zip](https://github.com/86Box/86Box/files/5742252/Philips.P3105.zip)

https://sites.google.com/site/atiwonderseriesdatabase/

I'm using MS-DOS 3.10 version, that does not boot in any other PC/XT machines.
But other versions seem to work right.

BUG: On 86box memory configuration 768kb can be specified, but system counts only 640kb.
I don't know which software/hardware can use the extra memory.

2. Columbia Data Systems 1600

Special hardware:

The most interesting of this machine is it has hard drive support on  its OEM MS-DOS 1.25, but these hard disk seem to be not standard 
and their size is 5mb, 7.5mb and 10mb. 

![Columbia Data Product hard dIsk](https://user-images.githubusercontent.com/31600563/103140143-d7e9fa80-46c1-11eb-9419-a0422598f3e8.png)

3. Zenith Data System Z-151/152/161

This machine cannot boot from a floppy disk if the CPU speed is 8 mhz or greater.
Boots fine at 4.77/7.16 mhz

Special hardware:

This model cames with MS-DOS 1.25 and 2.11. But the 2.11 is not 'pure' because its IO.SYS version is 1.08.
It also has includes a 'PART' utility instead the standard FDISK to crete partitions but it spects to be
a Winchester controller.

![Zenith Data Ssystem Z-151](https://user-images.githubusercontent.com/31600563/103140148-e9cb9d80-46c1-11eb-9941-af41ce3f432f.png)

![Zenith Z-151 hard diskpng](https://user-images.githubusercontent.com/31600563/103140149-ecc68e00-46c1-11eb-932e-c02aaf5b9e7d.png)

4. Zenith Data System Z-159

It uses OEM DOS 3.21. This version too, lacks the FDISK utility but I managed to install DOS to hard disk with 'PART' 
command since it work with generic hard disk controlers.

Special hardware:

This system also supports 1.2mb of RAM. On real machine, if  there is 1.2mb installed, it uses 640kb conventional and the rest is used as EMS memory.
This DOS version includes a EMM.SYS driver for the Zenith machines with expanded memory installed.

5. Zenith Data Sport Z-184

Special hardware:

It can have two 720kb floppy disk or an 1 FDD and one 20mb HDD. Similar to the Toshiba T1200. I don't know if the hard
disk is proprietary, however, one of the emulated hard disk controllers can be used.

This uses a LCD blue backlight screen like the T1200. 

It support a proprietary 1mb EMS RAM memory card. 

Here is the manual I found:
http://www.minuszerodegrees.net/manuals/Zenith%20Data%20Systems/ZDS%20SupersPort%20-%20Service%20Manual.pdf

**Additional context**
Some machines launched before 1986 has non-standard DOS versions. Some OEM 2.11 versions have MS-DOS 1.x components,
The reason for this, I think, it is the memory constrains of the computer of the period. Some of them were sold with only 128kb of RAM.
The other case are OEM DOS 3.0 and 3.1. The PC-DOS 3.0 and 3.1 does not boot on XT machines, only on AT. So I think some vendors modified these DOS versions to make compatible with their specific XT machines.
But they do not work on regular XT machines, like the Tandy 1000 DOS.

**Comment by OBattler** [comment]
Is there any technical documentation of these features?

**Comment by Fenix770** [comment]
I think this may be useful
For the Zenith 151:

I found the data sheet of, what seems to be, the Winchester hard disk. The manual is dated 1982,  the same year that MS-DOS 1.25 was launched, so It think it could be the HD controller that Zenith-151 expects to find.
Here is the link:
[http://www.bitsavers.org/pdf/codata/05-0003-01_Codata_Winchester_Disk_Controller_Manual_Oct82.pdf](url)

For the Columbia Data Products MPC-1600 

It seems to use a Winchester controller.
[http://www.minuszerodegrees.net/manuals/Columbia%20Data%20Products/Columbia%20Data%20Products%20MPC%201600%20-%20Operations%20Guide.pdf](url)

According another site it uses a Apex hard disk:
[http://www.vcfed.org/forum/showthread.php?1467-Columbia-1600-4-computer-with-hard-drive](url)

According stason.org the Apex HDs has capacities from 5mb to 20mb. These disk geometries are similar to the available options on its OEM DOS 1.25.
[https://stason.org/TULARC/pc/hard-drives-hdd/ampex/PYXIS-13-10MB-5-25-FH-MFM-ST412.html](url).

As for the EMS chipset I have not found anything yet. Same for the ATI card.

**Comment by OBattler** [comment]
OK, I'm going to tag this as v4.0 because that's going to be the Storage-focused version, so we can look into this stuff there.

Edit: Looks like it was tagged that already.
Edit #2: And by me, apparently. Evidently I forgot.

**Comment by EngiNerd89** [comment]
Just some side notes on the features requested:
- Philips P3105 video card: as this machine has no integrated video card. As, the ATI Small Wonder is an ISA card, this feature request is not specific to this machine.
- Philips P3105 640-768KB memory: it should be mapped at address d0000-effff and made available as EMS. I tried to fool the BIOS by making it available with mem_mapping_add, but it was not detected. 
- Zenith Z-151 clock issues: the real machine came only with a 4.77 MHz CPU, so floppy drive issues may happen there as well. Instead, Z-157/158/159 came with an 8 MHz CPU and with a different BIOS.

**Comment by Fenix770** [comment]
About P3105:

1. I didn't know the ATI card was not integrated as video chip. 
(Like the Amstrad 2086 VGA for example)

2. I suspect the extra memory in 640k-1024k on XT machines  not EMS or UMB.
One Example is the Juko XT.
I don't have  the data sheet of that machine, but I found specific device drivers that allow the use of the last 384kb of RAM as
disk cache or RAM drive. However, It cannot be used as EMS memory and cannot be mapped as UMB, to load device drivers either. 
3. As for the Z-159 a site at stason.org says it supports 1.2mb of RAM. Memory above 640kb is EMS, it includes a EMM.SYS device driver for Zenith DOS, like the  Toshiba MS-DOS on the T1000 laptop, but of course is a different hardware implementation. Unfortunately, I don't know if the documentation of the Z-159 exists. 
 
Here is the link:[https://stason.org/TULARC/pc/motherboards/Z/ZENITH-DATA-SYSTEMS-8088-Z-159.html](url)

**Comment by ghost** [comment]
I found brochures in website only specs. support to 1.2mb in mainboard? 
ems ram can be expanded to 5mb? edit: 
sorry for necropost 
src:
https://www.1000bit.it/js/web/viewer.html?file=%2Fad%2Fbro%2Fzenith%2Fzenithpc%2Dz159bcol%2Epdf#zoom=page-fit
https://www.1000bit.it/js/web/viewer.html?file=%2Fad%2Fbro%2Fzenith%2Fzenithpc%2Dz159bw%2Epdf#zoom=page-fit

**Comment by Fenix770** [comment]
There are other XT machines that can be expanded beyond 1m. Such as the Toshiba T1000LE can be expanded to 8Mb of LIM EMS RAM, and have a 1.44mb floppy drive something not common for an 8086 machine.
https://en.wikipedia.org/wiki/Toshiba_T1000LE.

I think the PS/2 model 30 8086 and the Amstrad 5086 both have the 82C100 chipset, which support up to 2.5mb of total RAM conventional+LIM EMS.
But the documentation of the hardware is required to implement it.

EMS memory can be used as RAM drive, disk cache, and DOS software such as Wordperfect, but I think 1mb of EMS is enough for an XT except when certain software such as C++ and Lotus 1-2-3 which can use more memory.
I think that 5mb of EMS is not so useful because apps like Windows 2.0 can only store up to 128kb of data on EMS memory, per apps. So is useful if load a lot of small apps. For bigger apps the Large page frame (LPF) is needed, LPF is a portion of conventional memory (usually the 256kb-640kb area) that can be bank-switched to the EMS memory, but that is included on some 286 chipsets and high end 16-bit ISA EMS boards.

Think that some old EMS boards like the Micromainfram 5150T, in the manual, recomends to add all memory chips on an 
specific order in order to work with the EMS driver (EMS32.BIN) the freeware driver EMS40.BIN does not require to put the chips in a particular order, but each time a data is required the board should find the particular, data chip per chip until the data is found, so the bigger is the program or database the longer is the time to find it.

Because these boards have usually only one hardware register, the LIM 4.0 can emulate extra registers in software, but is slower than the high end cards with several hardware register sets.

**Comment by lifengchun** [comment]
> 2\. I suspect the extra memory in 640k-1024k on XT machines  not EMS or UMB.
>         One Example is the Juko XT.
>         I don't have  the data sheet of that machine, but I found specific device drivers that allow the use of the last 384kb of RAM as
>         disk cache or RAM drive. However, It cannot be used as EMS memory and cannot be mapped as UMB, to load device drivers either.

To assist in the development of the [Juko XT](https://www.ultimateretro.net/en/motherboards/976) drivers, I made an extended memory patch [juko_pcxt_ext_mem.zip](https://github.com/86Box/86Box/files/9167806/juko_pcxt_ext_mem.zip) for the 86Box Juko XT that can use 640~1024KB of memory (verified by cdisk.sys from juko driver floppy and my jkhimem.sys).

**Comment by Fenix770** [comment]
Hello. 
Thanks for the patch. The jkhimem.sys driver you made, how it works? As far as I know the regular DOS HIMEM.SYS device driver is for 286 and later machines that have more than 1mb of RAM or can remap the last 384kb as extended memory. But XTs, don't support the use of XMS, because of the 1mb limit. So It can be used only as EMS if a chipset is present and accessed throughout the 64k page frame (the memory is outside the 1mb address space and usually mapped in D000-DFFF window region), or is used by a non-standard driver, that can be used as RAM disk, disk cache or the print spooler.

**Comment by lifengchun** [comment]
> The jkhimem.sys driver you made, how it works?

Yes, the function of the Juko XT chipset is relatively simple, and it can only map the high 384KB memory as a whole and fixedly to the low 0x2000 segment, so it cannot efficiently implement the EMS specification.
My jkhimem.sys implements the basic XMS 2.0 API (excluding lock, unlock and HMA). The method is to write 1 to the 0xe0 port to make the chipset map the high 384KB memory to the 0x2000 to 0x8000 segment, and then move the data according to the XMS handle. Because 0x2000-0x8000 segment occupies part of the basic memory, the moving process may require a temporary buffer, so the efficiency is relatively low.

**Comment by Cacodemon345** [comment]
The Winchester datasheets linked here do not appear to relate to the IBM PC architecture. However, I see two Western Digital Winchester controllers for it.

Which ones does these machines use, to be precise?

**Comment by Fenix770** [comment]
Sorry, I don't know, but these HDDs geometries predate the IBM standard, If I recall correctly the Partition software runes even DOS 1.25. The first IBM HDD was launched on October 1983 and was the type 1 10mb disk with the first revision of the Xebec controller.

**Comment by flama12333** [comment]
Columbia Data Products 1600 uses xebec hdd disk and the dump is uploaded by SomeGuy. sadly the dump look like bad
https://forum.vcfed.org/index.php?threads/honeywell-branded-columbia-data-products-1600.78062/#post-1360375
string :
(C) COPYRIGHT CDP 1983
